File: lib/typescript/examples/07-self-service-portal.ts
/**
 * Self-service portal backend: user fetches their own peer list and config.
 *
 * Real case: an Express/Fastify route that lets an authenticated user
 * download their own wg-quick config or QR code without admin involvement.
 * Uses provisioning endpoints which scope to the calling user's identity.
 */

import { WgPortalClient, WgPortalError } from "../src";

// In a real app these would come from the session / JWT
const LOGGED_IN_USER_EMAIL = "alice@corp.com";

const client = new WgPortalClient({
  baseUrl: "http://192.168.1.100:8888",
  username: "admin@wgportal.local",
  apiToken: process.env.WG_API_TOKEN!,
});

async function getSelfServiceInfo(email: string) {
  // Get the user's peer summary (works without knowing their userId)
  const info = await client.provisioning.getUserInfo({ email });

  console.log(`User: ${info.UserIdentifier}`);
  console.log(`Total peers: ${info.PeerCount}`);

  for (const peer of info.Peers) {
    console.log(`\n  Peer: ${peer.DisplayName ?? peer.Identifier}`);
    console.log(`  Interface: ${peer.InterfaceIdentifier}`);
    console.log(`  IPs: ${peer.IpAddresses?.join(", ")}`);
    console.log(`  Disabled: ${peer.IsDisabled ?? false}`);

    if (!peer.IsDisabled) {
      // Download config on demand
      const config = await client.provisioning.getPeerConfig(peer.Identifier);
      console.log(`\n  wg-quick config:\n${config}`);
    }
  }
}

getSelfServiceInfo(LOGGED_IN_USER_EMAIL).catch((err) => {
  if (err instanceof WgPortalError) {
    console.error(`${err.code}: ${err.message}`);
  } else {
    throw err;
  }
});

File: lib/rust/examples/07_self_service_portal.rs
/// 07_self_service_portal — Backend route: user fetches their own peer list and config by email.
///
/// Usage:
///   WG_BASE_URL=http://... WG_ADMIN_USER=... WG_API_TOKEN=... cargo run --example 07_self_service_portal <email>
use std::env;

use serde::Serialize;
use wg_portal_client::{WgPortalClient, WgPortalError};

fn require_env(key: &str) -> String {
    env::var(key).unwrap_or_else(|_| {
        eprintln!("Missing required env var: {key}");
        std::process::exit(1);
    })
}

#[derive(Serialize)]
struct PeerData {
    identifier: String,
    display_name: Option<String>,
    interface: String,
    ips: Option<Vec<String>>,
    disabled: bool,
    config: Option<String>,
}

#[derive(Serialize)]
struct UserPortal {
    user_identifier: String,
    peer_count: i64,
    peers: Vec<PeerData>,
}

fn get_user_peers(client: &WgPortalClient, email: &str) -> Result<UserPortal, WgPortalError> {
    let info = client.provisioning().get_user_info(None, Some(email))?;

    let mut peers_output = Vec::new();
    for p in &info.peers {
        let disabled = p.is_disabled.unwrap_or(false);
        let config = if !disabled {
            client
                .provisioning()
                .get_peer_config(&p.identifier)
                .ok()
        } else {
            None
        };

        peers_output.push(PeerData {
            identifier: p.identifier.clone(),
            display_name: p.display_name.clone(),
            interface: p.interface_identifier.clone().unwrap_or_default(),
            ips: p.ip_addresses.clone(),
            disabled,
            config,
        });
    }

    Ok(UserPortal {
        user_identifier: info.user_identifier,
        peer_count: info.peer_count,
        peers: peers_output,
    })
}

fn main() {
    let client = WgPortalClient::new(
        &require_env("WG_BASE_URL"),
        &require_env("WG_ADMIN_USER"),
        &require_env("WG_API_TOKEN"),
    );

    let args: Vec<String> = env::args().collect();
    let email = args.get(1).map(|s| s.as_str()).unwrap_or("alice@example.com");

    match get_user_peers(&client, email) {
        Ok(portal) => println!("{}", serde_json::to_string_pretty(&portal).unwrap()),
        Err(e) => eprintln!("Error [{code}]: {message}", code = e.code, message = e.message),
    }
}
